File: app/src/main/java/org/dhis2/usescases/programEventDetail/ProgramEventDetailPresenter.java
package org.dhis2.usescases.programEventDetail;

import androidx.annotation.NonNull;

import org.dhis2.commons.filters.data.FilterPresenter;
import org.dhis2.commons.filters.data.FilterRepository;
import org.dhis2.commons.schedulers.SchedulerProvider;
import org.dhis2.utils.analytics.matomo.MatomoAnalyticsController;
import org.dhis2.commons.filters.DisableHomeFiltersFromSettingsApp;
import org.dhis2.commons.filters.FilterItem;
import org.dhis2.commons.filters.FilterManager;
import org.dhis2.commons.filters.workingLists.EventFilterToWorkingListItemMapper;
import org.dhis2.commons.filters.workingLists.WorkingListItem;
import org.hisp.dhis.android.core.common.FeatureType;
import org.hisp.dhis.android.core.common.Unit;
import org.hisp.dhis.android.core.organisationunit.OrganisationUnit;
import org.hisp.dhis.android.core.program.Program;

import java.util.List;

import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.processors.FlowableProcessor;
import io.reactivex.processors.PublishProcessor;
import timber.log.Timber;

import static org.dhis2.utils.analytics.matomo.Actions.OPEN_ANALYTICS;
import static org.dhis2.utils.analytics.matomo.Actions.SYNC_EVENT;
import static org.dhis2.utils.analytics.matomo.Categories.EVENT_LIST;
import static org.dhis2.utils.analytics.matomo.Labels.CLICK;

public class ProgramEventDetailPresenter implements ProgramEventDetailContract.Presenter {

    private final ProgramEventDetailRepository eventRepository;
    private final SchedulerProvider schedulerProvider;
    private final FilterManager filterManager;
    private final FilterRepository filterRepository;
    private ProgramEventDetailContract.View view;
    CompositeDisposable compositeDisposable;
    private FlowableProcessor<Unit> listDataProcessor;
    private EventFilterToWorkingListItemMapper workingListMapper;
    private DisableHomeFiltersFromSettingsApp disableHomFilters;
    private MatomoAnalyticsController matomoAnalyticsController;

    public ProgramEventDetailPresenter(
            ProgramEventDetailContract.View view,
            @NonNull ProgramEventDetailRepository programEventDetailRepository,
            SchedulerProvider schedulerProvider,
            FilterManager filterManager,
            EventFilterToWorkingListItemMapper workingListMapper,
            FilterRepository filterRepository,
            FilterPresenter filterPresenter,
            DisableHomeFiltersFromSettingsApp disableHomFilters,
            MatomoAnalyticsController matomoAnalyticsController) {
        this.view = view;
        this.eventRepository = programEventDetailRepository;
        this.schedulerProvider = schedulerProvider;
        this.filterManager = filterManager;
        this.workingListMapper = workingListMapper;
        this.filterRepository = filterRepository;
        this.disableHomFilters = disableHomFilters;
        this.matomoAnalyticsController = matomoAnalyticsController;
        compositeDisposable = new CompositeDisposable();
        listDataProcessor = PublishProcessor.create();
    }

    @Override
    public void init() {
        compositeDisposable.add(
                Observable.just(filterRepository.programFilters(getProgram().uid()))
                        .subscribeOn(schedulerProvider.io())
                        .observeOn(schedulerProvider.ui())
                        .subscribe(
                                filters -> {
                                    if (filters.isEmpty()) {
                                        view.hideFilters();
                                    } else {
                                        view.setFilterItems(filters);
                                    }
                                },
                                Timber::e
                        )
        );

        compositeDisposable.add(FilterManager.getInstance().getCatComboRequest()
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(
                        catComboUid -> view.showCatOptComboDialog(catComboUid),
                        Timber::e
                )
        );

        compositeDisposable.add(Single.zip(
                Single.just(eventRepository.getAccessDataWrite()),
                eventRepository.hasAccessToAllCatOptions(),
                (hasWritePermission, hasAccessToAllCatOptions) ->
                        hasWritePermission && hasAccessToAllCatOptions)
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(
                        view::setWritePermission,
                        Timber::e)
        );

        compositeDisposable.add(
                eventRepository.program()
                        .observeOn(schedulerProvider.ui())
                        .subscribeOn(schedulerProvider.io())
                        .subscribe(
                                view::setProgram,
                                Timber::e
                        )
        );

        compositeDisposable.add(
                filterManager.ouTreeFlowable()
                        .subscribeOn(schedulerProvider.io())
                        .observeOn(schedulerProvider.ui())
                        .subscribe(
                                open -> view.openOrgUnitTreeSelector(),
                                Timber::e
                        )
        );

        compositeDisposable.add(
                filterManager.asFlowable().onBackpressureLatest()
                        .doOnNext(filterManager -> view.showFilterProgress())
                        .subscribeOn(schedulerProvider.io())
                        .observeOn(schedulerProvider.ui())
                        .subscribe(
                                filterManager -> view.updateFilters(filterManager.getTotalFilters()),
                                Timber::e
                        )
        );

        compositeDisposable.add(
                filterManager.getPeriodRequest()
                        .subscribeOn(schedulerProvider.io())
                        .observeOn(schedulerProvider.ui())
                        .subscribe(
                                periodRequest -> view.showPeriodRequest(periodRequest.getFirst()),
                                Timber::e
                        ));
    }

    @Override
    public void onSyncIconClick(String uid) {
        matomoAnalyticsController.trackEvent(EVENT_LIST, SYNC_EVENT, CLICK);
        view.showSyncDialog(uid);
    }

    public void addEvent() {
        view.startNewEvent();
    }

    @Override
    public void onBackClick() {
        view.back();
    }

    @Override
    public void onDettach() {
        compositeDisposable.clear();
    }

    @Override
    public void displayMessage(String message) {
        view.displayMessage(message);
    }

    @Override
    public void showFilter() {
        view.showHideFilter();
    }

    @Override
    public void clearFilterClick() {
        filterManager.clearAllFilters();
    }

    @Override
    public void filterCatOptCombo(String selectedCatOptionCombo) {
        FilterManager.getInstance().addCatOptCombo(
                eventRepository.getCatOptCombo(selectedCatOptionCombo)
        );
    }

    @Override
    public Program getProgram() {
        return eventRepository.program().blockingFirst();
    }

    @Override
    public FeatureType getFeatureType() {
        return eventRepository.featureType().blockingGet();
    }

    @Override
    public List<WorkingListItem> workingLists() {
        return eventRepository.workingLists().toFlowable()
                .flatMapIterable(data -> data)
                .map(eventFilter -> workingListMapper.map(eventFilter))
                .toList().blockingGet();
    }

    @Override
    public void clearOtherFiltersIfWebAppIsConfig() {
        List<FilterItem> filters = filterRepository.homeFilters();
        disableHomFilters.execute(filters);
    }

    @Override
    public void setOpeningFilterToNone(){
        filterRepository.collapseAllFilters();
    }

    @Override
    public String getStageUid() {
        return eventRepository.programStage().blockingGet().uid();
    }

    @Override
    public void setOrgUnitFilters(List<OrganisationUnit> selectedOrgUnits) {
        FilterManager.getInstance().addOrgUnits(selectedOrgUnits);
    }

    @Override
    public void trackEventProgramAnalytics() {
        matomoAnalyticsController.trackEvent(EVENT_LIST, OPEN_ANALYTICS, CLICK);
    }
}

File: app/src/main/java/org/dhis2/usescases/programEventDetail/ProgramEventDetailPresenter.kt
package org.dhis2.usescases.programEventDetail

import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.disposables.CompositeDisposable
import org.dhis2.commons.filters.DisableHomeFiltersFromSettingsApp
import org.dhis2.commons.filters.FilterManager
import org.dhis2.commons.filters.FilterManager.PeriodRequest
import org.dhis2.commons.filters.Filters
import org.dhis2.commons.filters.data.FilterRepository
import org.dhis2.commons.filters.workingLists.EventFilterToWorkingListItemMapper
import org.dhis2.commons.filters.workingLists.WorkingListItem
import org.dhis2.commons.matomo.Actions
import org.dhis2.commons.matomo.Categories
import org.dhis2.commons.matomo.Labels
import org.dhis2.commons.matomo.MatomoAnalyticsController
import org.dhis2.commons.schedulers.SchedulerProvider
import org.hisp.dhis.android.core.common.FeatureType
import org.hisp.dhis.android.core.organisationunit.OrganisationUnit
import org.hisp.dhis.android.core.program.Program
import timber.log.Timber

class ProgramEventDetailPresenter(
    private val view: ProgramEventDetailView,
    private val eventRepository: ProgramEventDetailRepository,
    private val schedulerProvider: SchedulerProvider,
    private val filterManager: FilterManager,
    private val workingListMapper: EventFilterToWorkingListItemMapper,
    private val filterRepository: FilterRepository,
    private val disableHomFilters: DisableHomeFiltersFromSettingsApp,
    private val matomoAnalyticsController: MatomoAnalyticsController,
) {
    val compositeDisposable: CompositeDisposable = CompositeDisposable()
    val program: Program?
        get() = eventRepository.program().blockingGet()
    val featureType: FeatureType?
        get() = eventRepository.featureType().blockingGet()
    val stageUid: String?
        get() = eventRepository.programStage().blockingGet()?.uid()

    fun init() {
        compositeDisposable.add(
            Observable.fromCallable {
                program?.uid()?.let { filterRepository.programFilters(it) } ?: emptyList()
            }
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(
                    { filters ->
                        if (filters.isEmpty()) {
                            view.hideFilters()
                        } else {
                            view.setFilterItems(filters)
                        }
                    },
                    { t -> Timber.e(t) },
                ),
        )
        compositeDisposable.add(
            FilterManager.getInstance().catComboRequest
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(
                    { catComboUid -> view.showCatOptComboDialog(catComboUid) },
                    { t -> Timber.e(t) },
                ),
        )
        compositeDisposable.add(
            Single.just(eventRepository.getAccessDataWrite())
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(
                    { aBoolean -> view.setWritePermission(aBoolean) },
                    { t -> Timber.e(t) },
                ),
        )
        compositeDisposable.add(
            eventRepository.program()
                .observeOn(schedulerProvider.ui())
                .subscribeOn(schedulerProvider.io())
                .subscribe(
                    { programModel -> view.setProgram(programModel!!) },
                    { t -> Timber.e(t) },
                ),
        )
        compositeDisposable.add(
            filterManager.ouTreeFlowable()
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(
                    { view.openOrgUnitTreeSelector() },
                    { t -> Timber.e(t) },
                ),
        )
        compositeDisposable.add(
            filterManager.asFlowable().onBackpressureLatest()
                .doOnNext { view.showFilterProgress() }
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(
                    { filterManager -> view.updateFilters(filterManager.totalFilters) },
                    { t -> Timber.e(t) },
                ),
        )
        compositeDisposable.add(
            filterManager.periodRequest
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(
                    { (first): Pair<PeriodRequest, Filters> -> view.showPeriodRequest(first) },
                    { t -> Timber.e(t) },
                ),
        )
    }

    fun onSyncIconClick(uid: String) {
        matomoAnalyticsController.trackEvent(
            Categories.EVENT_LIST,
            Actions.SYNC_EVENT,
            Labels.CLICK,
        )
        view.showSyncDialog(uid)
    }

    fun addEvent() {
        view.startNewEvent()
    }

    fun onBackClick() {
        view.back()
    }

    fun onDettach() {
        compositeDisposable.clear()
    }

    fun displayMessage(message: String) {
        view.displayMessage(message)
    }

    fun showFilter() {
        view.showHideFilter()
    }

    fun clearFilterClick() {
        filterManager.clearAllFilters()
    }

    fun filterCatOptCombo(selectedCatOptionCombo: String) {
        FilterManager.getInstance()
            .addCatOptCombo(eventRepository.getCatOptCombo(selectedCatOptionCombo))
    }

    fun workingLists(): List<WorkingListItem> {
        return eventRepository.workingLists().toFlowable()
            .flatMapIterable { data -> data }
            .map { eventFilter ->
                workingListMapper.map(eventFilter)
            }
            .toList().blockingGet()
    }

    fun clearOtherFiltersIfWebAppIsConfig() {
        val filters = filterRepository.homeFilters()
        disableHomFilters.execute(filters)
    }

    fun setOpeningFilterToNone() {
        filterRepository.collapseAllFilters()
    }

    fun setOrgUnitFilters(selectedOrgUnits: List<OrganisationUnit>) {
        FilterManager.getInstance().addOrgUnits(selectedOrgUnits)
    }

    fun trackEventProgramAnalytics() {
        matomoAnalyticsController.trackEvent(
            Categories.EVENT_LIST,
            Actions.OPEN_ANALYTICS,
            Labels.CLICK,
        )
    }

    fun trackEventProgramMap() {
        matomoAnalyticsController.trackEvent(
            Categories.EVENT_LIST,
            Actions.MAP_VISUALIZATION,
            Labels.CLICK,
        )
    }
}
